Helper/Discount.php

```diff
@@ -0,0 +1,320 @@
+        }
+
+        if ($accountModel->canApplyCardForQuote($quote)) {
+            $quoteGiftCard = $this->amastyQuoteFactory->getInstance()->create();
+            $this->amastyQuoteResource->getInstance()->load($quoteGiftCard, $quoteId, 'quote_id');
+            $subtotal = $quoteGiftCard->getSubtotal($quote);
+
+            if ($quoteGiftCard->getCodeId() && $accountModel->getCodeId() == $quoteGiftCard->getCodeId()) {
+
+                throw new LocalizedException(__('This gift card account is already in the quote.'));
+
+            } elseif ($quoteGiftCard->getGiftAmount() && $subtotal == $quoteGiftCard->getGiftAmount()) {
+
+                throw new LocalizedException(__('Gift card can\'t be applied. Maximum discount reached.'));
+
+            } else {
+                $quoteGiftCard->unsetData($quoteGiftCard->getIdFieldName());
+                $quoteGiftCard->setQuoteId($quoteId);
+                $quoteGiftCard->setCodeId($accountModel->getCodeId());
+                $quoteGiftCard->setAccountId($accountModel->getId());
+
+                $this->amastyQuoteRepository->getInstance()->save($quoteGiftCard);
+                $this->updateTotals($quote);
+
+                if ($this->getAmastyPayForEverything()) {
+                    // pay for everything, items, shipping, tax
+                    return $giftAmount;
+                } else {
+                    // pay for items only
+                    $totals = $quote->getTotals();
+                    return $totals[self::AMASTY_GIFTCARD]->getValue();
+                }
+            }
+        } else {
+            throw new LocalizedException(__('Gift card can\'t be applied.'));
+        }
+    }
+
+    /**
+     * Try to copy Amasty Gift Cart data from parent to immutable quote
+     *
+     * @param int|string $parentQuoteId
+     * @param int|string $immutableQuoteId
+     */
+    public function cloneAmastyGiftCards($parentQuoteId, $immutableQuoteId)
+    {
+        try {
+            $connection = $this->resource->getConnection();
+            $giftCardTable = $this->resource->getTableName('amasty_amgiftcard_quote');
+
+            // Clear previously applied gift cart codes from the immutable quote
+            $sql = "DELETE FROM {$giftCardTable} WHERE quote_id = :quote_id";
+            $connection->query($sql, ['quote_id' => $immutableQuoteId]);
+
+            // Copy all gift cart codes applied to the parent quote to the immutable quote
+            $sql = "INSERT INTO {$giftCardTable} (quote_id, code_id, account_id, base_gift_amount, code) 
+                    SELECT :immutable_quote_id, code_id, account_id, base_gift_amount, code 
+                    FROM {$giftCardTable} WHERE quote_id = :quote_id";
+            $connection->query($sql, ['immutable_quote_id' => $immutableQuoteId, 'quote_id' => $parentQuoteId]);
+        } catch (\Zend_Db_Statement_Exception $e) {
+            // The amasty_amgiftcard_quote table is missing, Amasty Gift Cart is not installed,
```

Comment:
can we check if amasty extension is there beginning of this function + early return, and here we track exception with bugsnag?

Reply:
@daisy1754  Sure.